File: my-folder/0032-longest-valid-parentheses/solution.py
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        stack = [-1]
        res = 0

        for i in range(len(s)):
            if s[i] == "(":
                # open
                stack.append(i)
            else:
                # close
                stack.pop()

                if len(stack) == 0: # change boundary marker
                    stack.append(i)
                else: # valid stack
                    res = max(res, i-stack[-1])

        return res

File: my-folder/0032-longest-valid-parentheses/solution.cpp
class Solution {
public:
    int longestValidParentheses(string s) {
        vector<int> stack = {-1};
        int res = 0;

        for (int i = 0; i < s.length(); i++) {
            if (s[i] == '(') {
                stack.push_back(i);
            } else {
                stack.pop_back();

                if (stack.size() == 0) {
                    stack.push_back(i); // new marker
                } else {
                    res = max(res, i - stack[stack.size() - 1]);
                }
            }
        }

        return res; 
    }
};
